docs/RFCS/version_migration.md

```diff
@@ -0,0 +1,245 @@
+- Feature Name: Version Migration
+- Status: draft
+- Start Date: 2017-07-10
+- Authors: Spencer Kimball
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue(s): N/A
+
+# Summary
+
+This RFC describes a mechanism for upgrading a CockroachDB cluster
+between point releases (e.g. from 1.0 to 1.1). In particular, it
+introduces the means to enforce version coherence amongst live nodes,
+which is necessary to gate access to new features. It further provides
+for a transition period during which all nodes settle on a new version
+*before* any are able to use new, backwards-incompatible features.
+During transition periods, the cluster may either be rolled back to
+the prior version, or else the operator may confirm the new version as
+the minimum version, allowing new, backwards-incompatible features to
+run.
+
+# Motivation
+
+Introducing backwards-incompatible features is currently risky, as
+usage could prevent orderly downgrades, or worse, could result in
+inconsistencies if used during a rolling upgrade.
+
+One example is the use of different storage for Raft's write ahead
+log
+[#16809](https://github.com/cockroachdb/cockroach/pull/16809). Having
+nodes begin using this new functionality means they cannot be
+downgraded to an earlier version of the binary without potential loss
+of data. To address this, CockroachDB migration must provide a
+transition period where all nodes upgrade to the version which
+supports the new Raft WAL storage, but do not yet use it. At this
+point, the cluster can still be downgraded to the prior version,
+allowing the operator to safely test the new build. When satisfied
+with the new version's stability, the operator confirms that the new
+version should become the minimum version, at which point the nodes
+begin using the Raft WAL storage, and the window for downgrading is
+closed.
+
+Another example is the `Revert` command. Once executed, it requires
+that all nodes performing reads understand how to utilize new
+information to ignore reverted data during reads. If one replica out
+of three were not upgraded to understand the effects of a `Revert`,
+then that node could incorrectly read reverted data as if it were
+still live.
+
+Backwards-incompatible changes to *existing* functionality are even
+more problematic because you'd expect them to be in active use at the
+time of upgrade. In this case, there would be no way to do a rolling
+upgrade without having some nodes with the old version and some nodes
+with the new, backwards-incompatible version, leading to potential
+inconsistencies.
+
+We expect most point releases to contain features requiring the
+migration support described in this RFC.
+
+# Detailed design
+
+Every node knows the version it's running, as the version is baked
+into the build. Nodes gossip a `NodeDescriptor`, which will be
+augmented to include the baked-in server version. Nodes also have a
```

Comment:
I think you want to say somewhere that the server itself has a baked in version and a minimum-supported-in-use-version